Add support for Date type in VectorHasher

aditi-pandit opened the pull request.

mbasmanova commented:
@aditi-pandit Aditi, there are build failures. Would you take a look?

aditi-pandit commented:
> @aditi-pandit Aditi, there are build failures. Would you take a look?

Hi Masha,

   I'm looking at the linux failures further.

  Just curious to know your thoughts on this change though.. Do you feel the solution works ?

mbasmanova commented on velox/exec/VectorHasher.h:
```diff
@@ -278,6 +284,21 @@ class VectorHasher {
+  }
+
+  template <>
+  inline int64_t hashData(bool value) const {
```
I wonder if this is needed. I think this can be changed to VELOX_UNREACHABLE() or removed.

mbasmanova commented on velox/exec/VectorHasher.h:
```diff
@@ -45,6 +45,11 @@ class UniqueValue {
   }
 
+  explicit UniqueValue(Date value) {
+    size_ = sizeof(Date);
```
I think this should be sizeof(int64_t). sizeof(Date) works by chance, since it happens to be equal to sizeof(int64_t).

aditi-pandit replied:
I don't fully understand the reasoning behind making it sizeof(int64_t). Please can you give more explanation.

mbasmanova replied:
size_ is expected to specify the number of valid bytes in data_. You are storing a int64_t (value.days()) in data_, hence, the size of that is sizeof(int64_t). It happens so that sizeof(Date) == sizeof(int64_t), but that's just a coincidence.

aditi-pandit replied:
I see. Thanks for the explanation.

mbasmanova reviewed (commented):
@aditi-pandit Looks good overall. Some comments below.

aditi-pandit commented:
Hi Masha,

Sorry for the delay on this. I wanted to get the tpch queries out of the way first.

I have addressed the review comments and fixed the linux compilation failure.

Please can you take a look.

mbasmanova reviewed (approved):
@aditi-pandit Looks good to me % comment about sizeof.

aditi-pandit commented:
Fixing the failure.

laithsakka commented on velox/exec/VectorHasher.h:
```diff
@@ -383,11 +397,11 @@ class VectorHasher {
     bool inRange = true;
     rows.template testSelected([&](vector_size_t row) {
-      if (values[row] > max_ || values[row] < min_) {
+      if (toInt64(values[row]) > max_ || toInt64(values[row]) < min_) {
```
nit:
auto something = toInt64(values[row])
... use something

laithsakka commented on velox/exec/VectorHasher.h:
```diff
@@ -398,18 +412,19 @@ class VectorHasher {
   uint64_t valueId(T value) {
     if (isRange_) {
-      if (value > max_ || value < min_) {
+      if (toInt64(value) > max_ || toInt64(value) < min_) {
```
ditto

laithsakka commented on velox/exec/VectorHasher.h:
```diff
@@ -419,10 +434,10 @@ class VectorHasher {
   uint64_t lookupValueId(T value) const {
     if (isRange_) {
-      if (value > max_ || value < min_) {
+      if (toInt64(value) > max_ || toInt64(value) < min_) {
```
ditto

laithsakka commented on velox/exec/tests/VectorHasherTest.cpp:
```diff
@@ -407,6 +407,55 @@ TEST_F(VectorHasherTest, integerIds) {
 
+TEST_F(VectorHasherTest, dateIds) {
+  auto vector = BaseVector::create(DATE(), 100, pool_.get());
+  auto dates = vector->as<FlatVector<Date>>();
```
nit: auto* dates

laithsakka commented on velox/exec/tests/VectorHasherTest.cpp:
```diff
@@ -407,6 +407,55 @@ TEST_F(VectorHasherTest, integerIds) {
+  hasher->enableValueIds(1, 100000);
+  // We add values that are over 100K distinct and with max - min > int32_t max.
+  EXPECT_TRUE(hasher->computeValueIds(*vector, rows, hashes));
+  // null is still 0.
```
nit: Hash of null is still 0.

laithsakka commented on velox/exec/tests/VectorHasherTest.cpp:
```diff
@@ -407,6 +407,55 @@ TEST_F(VectorHasherTest, integerIds) {
+  EXPECT_FALSE(hasher->computeValueIds(*vector, rows, hashes));
+  hasher->enableValueRange(1, 2000);
+  EXPECT_TRUE(hasher->computeValueIds(*vector, rows, hashes));
+  // null is always 0
```
nit: Hash of null is always 0.

laithsakka commented on velox/exec/tests/VectorHasherTest.cpp:
```diff
@@ -407,6 +407,55 @@ TEST_F(VectorHasherTest, integerIds) {
+  EXPECT_TRUE(hasher->computeValueIds(*vector, rows, hashes));
+  // null is still 0.
+  EXPECT_EQ(hashes[0], 0);
+  for (int count = 0; count < 1000; ++count) {
```
nit: for (auto count

laithsakka commented on velox/exec/VectorHasher.h:
```diff
@@ -45,6 +45,13 @@ class UniqueValue {
 
+  explicit UniqueValue(Date value) {
+    // The number of valid bytes of Date stored in data_ is
+    // int64_t(value.days()).
```
what does  int64_t(value.days()) means?
